src/daft-core/src/array/ops/utf8.rs

```diff
@@ -787,6 +810,195 @@ impl Utf8Array {
         Ok(result)
     }
 
+    pub fn substr<I, U>(
+        &self,
+        start: &DataArray<I>,
+        length: Option<&DataArray<U>>,
+    ) -> DaftResult<Utf8Array>
+    where
+        I: DaftIntegerType,
+        <I as DaftNumericType>::Native: Ord,
+        U: DaftIntegerType,
+        <U as DaftNumericType>::Native: Ord,
+    {
+        let (is_full_null, expected_size) = parse_inputs(self, &[start])
+            .map_err(|e| DaftError::ValueError(format!("Error in substr: {e}")))?;
+
+        if is_full_null {
+            return Ok(Utf8Array::full_null(
+                self.name(),
+                &DataType::Utf8,
+                expected_size,
+            ));
+        }
+
+        let self_iter = create_broadcasted_str_iter(self, expected_size);
+
+        let result: Utf8Array = match length {
+            Some(length) => {
+                if length.len() != 1 && length.len() != expected_size {
+                    return Err(DaftError::ValueError(
+                        "Inputs have invalid lengths: length".to_string(),
+                    ));
+                }
+
+                match (start.len(), length.len()) {
```

Comment:
Instead of matching all these cases, you could look into using `iter::repeat` and `take`, like `iter::repeat(start.get(0).unwrap()).take(expected_size)`

Reply:
I tried looking into this at first, but found two problems I couldn't get around:

1. `start.as_arrow().iter()` return value of type `ZipValidity<&<I as DaftNumericType>, Iter<<...>, BitmapIter>`. I didn't manage to create an iterator of a similar type with `repeat`.
2. From the performance point of view, if we will have iterators in both cases to reduce `match` arms, we most likely will need to do a type check in `map` for every value, even though it does not make sense for iterator created with `repeat`

```rust
  .map(|(val, s)| match (val, s) {
      (Some(val), Some(s)) => {
          // This type check will be required for both variables start and length for all iterations
          let s: usize = NumCast::from(*s).ok_or_else(|| {
              DaftError::ComputeError(format!(
                  "Error in repeat: failed to cast start as usize {s}"
              ))
          })?;
          Ok(substring(val, s, length)
```

Reply:
Try to take a stab at it, but if it gets too complicated, don't worry about it

Reply:
Hey @kevinzwang, I actually happen to have time to try that out today. 
It turned out to be much more compact this way once I finally figured out all the types of generics and compatible iterators.  

Take a look at the PR once you have the time 🙇‍♂️